src/OpenApi/src/OpenApiGenerator.cs

```diff
@@ -0,0 +1,474 @@
+/// </summary>
+internal class OpenApiGenerator
+{
+    private readonly IHostEnvironment? _environment;
+    private readonly IServiceProviderIsService? _serviceProviderIsService;
+
+    private static readonly ParameterBindingMethodCache ParameterBindingMethodCache = new();
+
+    /// <summary>
+    /// Creates an <see cref="OpenApiGenerator" /> instance given a <see cref="IHostEnvironment" />
+    /// and a <see cref="IServiceProviderIsService" /> instance.
+    /// </summary>
+    /// <param name="environment">The host environment.</param>
+    /// <param name="serviceProviderIsService">The service to determine if the a type is available from the <see cref="IServiceProvider"/>.</param>
+    internal OpenApiGenerator(
+        IHostEnvironment? environment,
+        IServiceProviderIsService? serviceProviderIsService)
+    {
+        _environment = environment;
+        _serviceProviderIsService = serviceProviderIsService;
+    }
+
+    /// <summary>
+    /// Generates an <see cref="OpenApiPathItem"/> for a given <see cref="Endpoint" />.
+    /// </summary>
+    /// <param name="methodInfo">The <see cref="MethodInfo"/> associated with the route handler of the endpoint.</param>
+    /// <param name="metadata">The endpoint <see cref="EndpointMetadataCollection"/>.</param>
+    /// <param name="pattern">The route pattern.</param>
+    /// <returns>An <see cref="OpenApiPathItem"/> annotation derived from the given inputs.</returns>
+    internal OpenApiOperation? GetOpenApiOperation(
+        MethodInfo methodInfo,
+        EndpointMetadataCollection metadata,
+        RoutePattern pattern)
+    {
+        if (metadata.GetMetadata<IHttpMethodMetadata>() is { } httpMethodMetadata &&
+            httpMethodMetadata.HttpMethods.SingleOrDefault() is { } method &&
+            metadata.GetMetadata<IExcludeFromDescriptionMetadata>() is null or { ExcludeFromDescription: false })
+        {
+            return GetOperation(method, methodInfo, metadata, pattern);
+        }
+
+        return null;
+    }
+
+    private OpenApiOperation GetOperation(string httpMethod, MethodInfo methodInfo, EndpointMetadataCollection metadata, RoutePattern pattern)
+    {
+        var disableInferredBody = ShouldDisableInferredBody(httpMethod);
+        return new OpenApiOperation
+        {
+            OperationId = metadata.GetMetadata<IEndpointNameMetadata>()?.EndpointName,
+            Summary = metadata.GetMetadata<IEndpointSummaryMetadata>()?.Summary,
+            Description = metadata.GetMetadata<IEndpointDescriptionMetadata>()?.Description,
+            Tags = GetOperationTags(methodInfo, metadata),
+            Parameters = GetOpenApiParameters(methodInfo, metadata, pattern, disableInferredBody),
+            RequestBody = GetOpenApiRequestBody(methodInfo, metadata, pattern),
+            Responses = GetOpenApiResponses(methodInfo, metadata)
+        };
+
+        static bool ShouldDisableInferredBody(string method)
```

Comment:
super nit: should this be in a shared location?

Reply:
Meh, considering how small it is I think it's almost not worth the effort of putting it in a shared location/adding includes/etc.